### PR TITLE
refactor particle source import/export

### PR DESCRIPTION
**Pull request by gridley**
This PR does two things

1) Refactor and generalize particle source import and export (whether via MCPL or HDF5)
2) ~~Adds information about whether OpenMC was compiled with ncrystal and mcpl to `openmc --version`'s output~~ (this was already done on wednesday)

The reason for 1) is that export and import of particle sources has applications beyond writing the fission source or surface source. We are doing some work on a dynamic MC implementation, and found the current facilities for writing sources to be a little lackluster. In particular, the previous function calls took as input a filename and a boolean stating whether the surface source bank is what should be written.

In this, the function calls for reading/writing sources are generalized. They now take the filename, a gsl::span<SourceSite> (in order to allow passing in values from either a SharedArray or vector), and a vector<int64_t> which gives the parallel scan of global source site indices across all ranks.

There used to be a function called compute_surface_source_site_indices that was very specific for calculating the global indices of particles on the surface source across MPI ranks. This was in fact doing a very basic MPI operation: pass an int64 in, and output a vector of the exclusive parallel scan of those ints across the ranks. Because this is useful for writing other types of source, it has been moved to `message_passing.h` and renamed accordingly.

On top of that, there was some duplicate code that handled automatically naming the source files. I've added a function to `file_utils.h` that extracts the file extension, and lets us automatically give the correct file extension if we're writing either MCPL files or HDF5 files.

The application of this is that Sebastian will soon be able to write out slices of the steady state particle population as described in [this Serpent wiki article](https://serpent.vtt.fi/mediawiki/index.php/Transient_simulations) (see "generating the steady state source distribution") with ease. We have a slightly different implementation that doesn't assume delta tracking but rather track segment time info (thanks @cfichtlscherer!). This more generalized code will be useful because we can use it for both the neutrons in flight and generating precursor banks when we opt to use pointwise precursors.

Also @myerspat, the Catch2 submodule is freakin sick! It's so nice for writing new C++ code. I've added some basic testing of the name file extension extractor.

One thing I need feedback on is a change I've made in `write_source_bank`. It seems to me that the previous behavior for surface sources may have not been correct if the surface source was not completely filled on each MPI rank. This variable called `count_size` should have been the number of sites to write on each rank, but it seems to have been using `work_per_rank` instead. It seems like this would potentially write out of bounds. I may add some Catch2 testing to figure out what we should actually have here, but am not sure how to test MPI stuff in Catch2. I don't completely understand how the parallel HDF5 code here is working. I'm guessing @paulromano  is the only one here familiar with how parallel HDF5 should be used?

### CODE REVIEW

**Comment by paulromano** [comment]
Thanks @gridley for this PR! Looking forward to reviewing this. Juts a heads up I'm over in Paris this week giving an OpenMC course so I might not be able to get to this until next week.

**Comment by gridley** [comment]
OK, thanks @paulromano; it's all good to let this sit for a bit.

**Comment by yrrepy** [comment]
Is is possible to get something in there to determine the source's minimum weight?

It would be beneficial to be able to renormalize weight and weight_avg for survival biasing

**Comment by gridley** [comment]
> Is is possible to get something in there to determine the source's minimum weight?
> 
> It would be beneficial to be able to renormalize weight and weight_avg for survival biasing

Well, that's very easy to do. Do you mean on a surface source? Or on a fission source? Yes, for sure, I could add in code to do this with a std::accumulate and MPI_Reduce call on a SourceSite bank.

**Comment by yrrepy** [comment]
> > Is is possible to get something in there to determine the source's minimum weight?
> > It would be beneficial to be able to renormalize weight and weight_avg for survival biasing
> 
> Well, that's very easy to do. Do you mean on a surface source? Or on a fission source? Yes, for sure, I could add in code to do this with a std::accumulate and MPI_Reduce call on a SourceSite bank.
> 
> If it's something you want done after running the simulation, it would probably be best to write some python function to extract the minimum particle weight from a source file.

It would be for any fixed source (surface or generic source (C++ lib, spatial))

It would be best to have handy in-code, when executing.
It's something I've been mulling. One point of weakness in OpenMC is for fixed sources there is no automatic renorm of survival biasing weight parameters.
In my own calculations, I do it by hand, scoping the source file first, then renorming weight and weight_avg in settings.xml
But eventually this is going to get to be pretty cumbersome for me.

https://permalink.lanl.gov/object/tr?what=info:lanl-repo/lareport/LA-UR-22-30006
pdf page 342, item 3  (verbiage on wcX<0)

It's would be a number than useable below, to renorm weight_cutoff  and weight_survive.
https://github.com/openmc-dev/openmc/blob/develop/src/settings.cpp

Which I could see to in a future PR.

**Comment by gridley** [comment]
What do you mean by renorming? What is the specific operation? I went to that page in the manual, and my PDF reader took me to the actual page 342. I only see some plot about cross section broadening.

Either way, I think this PR would not really be the place to put it. We'd probably just want to add some extra setting to run this initial setup after reading the source from a file. I'd be happy to put this in for you, because it should be quite straightforward.

**Comment by yrrepy** [comment]
actual page 338

renorming: multiplying weight and weight_avg by the minimum weight on the fixed source

yeah, the whole affair should happen in another PR, just seemed opportune to try and hop in here and get that first needed part (in-code determination of the min. source weight)

File: src/message_passing.cpp [hunk 1]
```diff
@@ -17,6 +17,29 @@ extern "C" bool openmc_master()
   return mpi::master;
 }
 
+vector<int64_t> calculate_parallel_index_vector(const int64_t size)
```
Comment (paulromano) [inline src/message_passing.cpp:20]:
```suggestion
vector<int64_t> calculate_parallel_index_vector(int64_t size)
```

File: src/message_passing.cpp [hunk 2]
```diff
@@ -17,6 +17,29 @@ extern "C" bool openmc_master()
...
+#ifdef OPENMC_MPI
+  result.resize(n_procs);
+  vector<int64_t> bank_size(n_procs);
+
+  // Populate the result with cumulative sum of the number of
+  // surface source banks per process
+  MPI_Scan(&size, bank_size.data(), 1, MPI_INT64_T, MPI_SUM, intracomm);
+  MPI_Allgather(
+    bank_size.data(), 1, MPI_INT64_T, result.data(), 1, MPI_INT64_T, intracomm);
+  result.insert(result.begin(), 0);
```
Comment (paulromano) [inline src/message_passing.cpp:34]: Inserting at position zero requires reallocating -- I think it'd be cleaner to `resize(n_procs + 1)` above and then `Allgather` into `result.data() + 1`.

File: src/output.cpp [hunk 3]
```diff
@@ -390,6 +390,12 @@ void print_build_info()
 #ifdef COVERAGEBUILD
   coverage = y;
 #endif
+#ifdef NCRYSTAL
+  ncrystal = y;
+#endif
+#ifdef OPENMC_MCPL
+  mcpl = y;
+#endif
```
Comment (paulromano) [inline src/output.cpp:398]:
```suggestion
```

File: src/state_point.cpp [hunk 4]
```diff
@@ -47,6 +49,15 @@ extern "C" int openmc_statepoint_write(const char* filename, bool* write_source)
       simulation::current_batch, w);
   }
 
+  // If a file name was specified, ensure it has .h5 file extension
+  const auto extension = get_file_extension(filename_);
+  if (extension == "") {
+    filename_.append(".h5");
+  } else if (extension != "h5") {
+    fatal_error("openmc_statepoint_write was passed an incorrect file "
+                "extension. Must either have no file extension or .h5");
```
Comment (paulromano) [inline src/state_point.cpp:58]: Does it really matter what the extension is? If someone wants to write a statepoint called "gavin.is.awesome", is there a good reason to stop them?

File: src/message_passing.cpp (hunk 2 above)
Comment (paulromano) [reply src/message_passing.cpp:34]: Ah sorry, just realizing this is not code that _you_ wrote and was just moved from elsewhere. Feel free to disregard, or update if you feel so inclined!

File: src/state_point.cpp [hunk 5]
```diff
@@ -577,58 +565,59 @@ void write_source_point(const char* filename, bool surf_source_bank)
...
+  // the old code was
+  //
+  // int64_t dims_size = settings::n_particles;
+  // int64_t count_size = simulation::work_per_rank;
+  //
+  // but that doesn't make sense! The count on a rank is not
+  // necessarily equal to work_per_rank. For the case of the
+  // fission source bank, that's true, but for surface source
+  // creation, there may be a different number of particles
+  // on each rank.
```
Comment (paulromano) [inline src/state_point.cpp:609]: If you look at the develop branch, although `count_size` starts off as being equal to `work_per_rank`, for the surface source case it gets set again to `surf_source_bank.size()`:
https://github.com/openmc-dev/openmc/blob/ae615c02d9bb97f05f52a6b74494f1dc37514186/src/state_point.cpp#L619-L622

File: tests/cpp_unit_tests/test_file_utils.cpp [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  // not sure how to test this when running on windows?
+  REQUIRE(dir_exists("/"));
+
+  // if this exists on your system... you deserve for this test to fail
+  REQUIRE(!dir_exists("/asdfa/asdfasdf/asdgasodgosuihasjkgh/"));
+}
+
+TEST_CASE("Test file_exists")
+{
+  // TODO make a file test it exists, delete it
```
Comment (paulromano) [inline tests/cpp_unit_tests/test_file_utils.cpp:27]: Were you planning on completing this one as part of this PR?

File: tests/cpp_unit_tests/test_file_utils.cpp [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+using namespace openmc;
+
+TEST_CASE("Test get_file_extension")
+{
+  REQUIRE(get_file_extension("rememberthealamo.png") == "png");
+  REQUIRE(get_file_extension("statepoint.20.h5") == "h5");
+  REQUIRE(get_file_extension("wEiRDNaa_ame.h4") == "h4");
+  REQUIRE(get_file_extension("has_directory/asdf.20.h5") == "h5");
+  REQUIRE(get_file_extension("wasssssup_lol") == "");
+  REQUIRE(get_file_extension("has_directory/secret_file") == "");
```
Comment (paulromano) [inline tests/cpp_unit_tests/test_file_utils.cpp:13]: Maybe one more you could add is "lovely.dir/extensionless_file"

**Review by paulromano — CHANGES_REQUESTED** [review]
Thanks for the PR @gridley! Nice to have someone else thinking about structural improvements :smile: A few small comments but I don't see any major issues here.

File: include/openmc/file_utils.h [hunk 8]
```diff
@@ -33,6 +35,30 @@ inline bool file_exists(const std::string& filename)
   return s.good();
 }
 
+// Gets the file extension of whatever string is passed in. This is defined as
+// a sequence of strictly alphanumeric characters which follow the last period,
+// i.e. at least one alphabet character is present, and zero or more numbers.
+// If such a sequence of characters is not found, an empty string is returned.
+inline std::string get_file_extension(const std::string& filename)
```
[the remark this answers is not in the document]
Comment (gridley) [reply include/openmc/file_utils.h:42]: Agreed. Will move to a source file.

File: include/openmc/state_point.h [hunk 9]
```diff
@@ -3,18 +3,44 @@
...
+// or fission source sites for eigenvalue continuation runs.
+//
+// This function ends up calling write_source_bank, and is responsible
+// for opening the file to be written to and controlling whether the
+// write is done in parallel (if compiled with parallel HDF5).
+//
+// bank_index is an exclusive parallel scan of the source_bank.size()
+// values on each rank, used to create global indexing. This vector
+// can be created by calling calculate_parallel_index_vector on
+// source_bank.size() if such a vector is not already available.
```
[the remark this answers is not in the document]
Comment (gridley) [reply include/openmc/state_point.h:35]: 🥲

File: tests/cpp_unit_tests/test_file_utils.cpp (hunk 7 above)
Comment (gridley) [reply tests/cpp_unit_tests/test_file_utils.cpp:13]: very good catch! It was not correct in that case. Should be fixed now.

File: src/state_point.cpp (hunk 4 above)
Comment (gridley) [reply src/state_point.cpp:58]: I agree, I've changed it to a helpful warning message instead saying that an hdf5 file will be written regardless of the extension passed in.

**Comment by gridley** [comment]
Alright, I believe all comments have now been addressed!

**Comment by gridley** [comment]
I'm not sure if the CI failure is due to my changes. It says that pybind11 is missing. Hmmmm..

File: tests/cpp_unit_tests/test_file_utils.cpp (hunk 7 above)
Comment (paulromano) [reply tests/cpp_unit_tests/test_file_utils.cpp:13]: testing ftw :smile:

File: src/state_point.cpp [hunk 10]
```diff
@@ -577,58 +565,59 @@ void write_source_point(const char* filename, bool surf_source_bank)
...
+  //
+  // Hence, I have changed count_size to be the number
+  // on this rank rather than work_per_rank. Similarly,
+  // dims_size used to be n_particles, but for surface sources,
+  // we are not guaranteed to create n_particles at the surface.
+  // As a result, I've changed this to be the total size of
+  // the bank across all ranks, which maintains the correct
+  // behavior for fission bank outputs.
+
```
Comment (paulromano) [inline src/state_point.cpp:618]: Can get rid of this now
```suggestion
```

File: src/state_point.cpp [hunk 11]
```diff
@@ -577,58 +565,59 @@ void write_source_point(const char* filename, bool surf_source_bank)
...
 
-    filename_ = fmt::format("{0}source.{1:0{2}}.h5", settings::path_output,
-      simulation::current_batch, w);
+  std::string filename_(filename);
+  const auto extension = get_file_extension(filename_);
+  if (extension == "") {
+    filename_.append(".h5");
+  } else if (extension != "h5") {
+    fatal_error("openmc_source_point was passed an incorrect file "
+                "extension. Must either have no file extension or .h5");
```
Comment (paulromano) [inline src/state_point.cpp:577]: Same consideration here -- let's make this a warning. Also `openmc_source_point` is not a thing, so the wording should be updated as well

**Review by paulromano — CHANGES_REQUESTED** [review]
Two small requests and then this should be good to go. Adding a commit on this branch should also resolve the CI failure you were seeing before.

**Comment by gridley** [comment]
Hey @paulromano, I'm a bit stumped as to why CI is failing. It seems that builds using MPI are hanging on a test called `statepoint_sourcesep`. I've tried running this on my system both with an MPI build that uses parallel HDF5 and does not use parallel HDF5, and both have the test passing instantly on my system. It is something with the HDF or MPI version being used on the CI server.

Do you have a recommendation on how to dig into this?

**Comment by paulromano** [comment]
Hmmm, that's a tough one. I don't have any great ideas here other than maybe to set up a vanilla Ubuntu 22.04 image on a cloud server and try to replicate the GitHub Actions installation to the best of your ability. I seem to recall @pshriwise mentioning some way to do better debugging on GHA?

**Comment by pshriwise** [comment]
Ramping up again after some paternity leave and just saw this. There's a really handy github action you can insert into the CI YAML file at the right location and it will provide you with an SSH link to the CI machine where you can do your own debugging of the build/tests. Hope it helps @gridley!

**Comment by gridley** [comment]
We appear to be good to go here now!

**Review by paulromano — APPROVED** [review]

**Comment by paulromano** [comment]
Thanks @gridley!

**Pull request merged**